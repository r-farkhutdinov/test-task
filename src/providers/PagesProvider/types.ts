import { ReactNode } from 'react';
import { TableData } from '../../types';

export type PagesContextValue = {
    data?: TableData;
    activePage?: string;
    query?: string;
    loading?: boolean;
    error?: string;
    changeActivePage?: (activePage: string) => void;
    setQuery?: (query: string) => void;
    load?: (q?: string) => void;
};

export type Props = {
    children: ReactNode;
}