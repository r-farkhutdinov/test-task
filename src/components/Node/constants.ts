export const basePadding = 22;
export const padding = 16;
export const nodeRootTestId = 'node-container';