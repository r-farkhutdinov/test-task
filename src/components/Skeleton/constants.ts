export const defaultLeftPadding = 16;
export const defaultRightPadding = 32;
export const skeletonRootTestId = 'skeleton';