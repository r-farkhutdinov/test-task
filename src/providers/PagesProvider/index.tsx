import { createContext, useCallback, useEffect, useMemo, useState } from 'react';
import { fetchData } from '../../api';
import type { TableData } from '../../types';
import type { PagesContextValue, Props } from './types';

export const PagesContext = createContext<PagesContextValue>({});

const errorMessage = 'Error fetching data. It is likely that the server is not running. Please start the server from its directory.';

/**
 * Context provider with the core logic of TOC: managing state, fetching data
 */
export const PagesProvider = ({ children }: Props) => {
    const [activePage, setActivePage] = useState<string>();
    const [data, setData] = useState<TableData>();
    const [query, setQuery] = useState('');
    const [error, setError] = useState<string>();
    const [loading, setLoading] = useState(false);

    const load = async (q?: string) => {
        setActivePage(undefined);
        try {
            setLoading(true);
            const result = await fetchData(q);
            setData(result);
        } catch {
            setError(errorMessage);
        } finally {
            setLoading(false);
        }
    };

    // this effect will fire twice in strict mode https://react.dev/reference/react/useEffect 
    useEffect(() => {
        load();
    }, []);

    const changeActivePage = useCallback((page: string) => {
        if (activePage !== page) {
            setActivePage(page);
        }
    }, [activePage]);

    const value = useMemo(() => ({
        activePage,
        data,
        query,
        loading,
        error,
        load,
        setQuery,
        changeActivePage
    }), [activePage, changeActivePage, data, error, loading, query]);

    return (
        <PagesContext.Provider value={value}>{children}</PagesContext.Provider>
    );
};