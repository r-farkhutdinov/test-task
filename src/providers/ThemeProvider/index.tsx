import { createContext, useCallback, useMemo, useState } from 'react';
import { storage } from '../../utils/storage';
import { type ThemeContextValue, type Props, Theme } from './types';

export const themeStorageKey = 'theme';

/**
 * Context provider with theme logic
 */
export const ThemeContext = createContext<ThemeContextValue>({
    theme: Theme.Light,
});

export const ThemeProvider = ({ children }: Props) => {
    const storedTheme = storage.get(themeStorageKey) as Theme;

    const [theme, setTheme] = useState(storedTheme ?? Theme.Light);

    const toggleTheme = useCallback(() => {
        const newTheme = theme === Theme.Light ? Theme.Dark : Theme.Light;
        setTheme(newTheme);
        storage.set(themeStorageKey, newTheme);
    }, [theme]);

    const value = useMemo(() => ({
        theme,
        toggleTheme
    }), [theme, toggleTheme]);

    return (
        <ThemeContext.Provider value={value}>{children}</ThemeContext.Provider>
    );
};