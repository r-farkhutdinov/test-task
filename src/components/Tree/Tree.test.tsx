import { render, screen } from '@testing-library/react';
import { PagesContext } from '../../providers/PagesProvider';
import type { PagesContextValue } from '../../providers/PagesProvider/types';
import type { TreeProps } from './types';
import { rootTreeTestId } from './constants';
import { Tree } from '.';

const mockData = {
    topLevelIds: ['Page'],
    entities: {
        pages: {}
    }
};

const defaultContextValue: PagesContextValue = { data: mockData, loading: false };

const renderTreeWithContext = (context: PagesContextValue, props?: TreeProps) => render(
    <PagesContext.Provider value={context}>
        <Tree {...props} />
    </PagesContext.Provider>
);

describe('Tree component', () => {
    it('matches snapshot without pages', () => {
        const { asFragment } = renderTreeWithContext(defaultContextValue);
        expect(asFragment()).toMatchSnapshot();
    });

    it('matches snapshot with pages and isActive', () => {
        const { asFragment } = renderTreeWithContext(defaultContextValue, { pages: ['Page 1', 'Page 2'], isActive: true });
        expect(asFragment()).toMatchSnapshot();
    });

    it('matches snapshot with pages, isActive, and isActiveLevelTwo', () => {
        const { asFragment } = renderTreeWithContext(defaultContextValue, { pages: ['Page 1', 'Page 2'], isActive: true, isActiveLevelTwo: true });

        expect(asFragment()).toMatchSnapshot();
    });

    it('should render correctly with pages', () => {
        renderTreeWithContext(defaultContextValue, { pages: ['Page 1', 'Page 2'], isActive: true });

        const tree = screen.getByTestId(rootTreeTestId);
        expect(tree).toBeInTheDocument();
    });
});
