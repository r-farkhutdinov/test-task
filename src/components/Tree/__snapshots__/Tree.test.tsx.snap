// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tree component matches snapshot with pages and isActive 1`] = `
<DocumentFragment>
  <div
    class="root active"
    data-testid="tree"
  />
</DocumentFragment>
`;

exports[`Tree component matches snapshot with pages, isActive, and isActiveLevelTwo 1`] = `
<DocumentFragment>
  <div
    class="root active activeLevelTwo"
    data-testid="tree"
  />
</DocumentFragment>
`;

exports[`Tree component matches snapshot without pages 1`] = `
<DocumentFragment>
  <div
    class="root"
    data-testid="tree"
  />
</DocumentFragment>
`;
